# 3. Generating expressed GRC data 

For a GRC-linked gene to be considered expressed in this study, it must satisfy two conditions
- Have an TPM > cutoff generated in /02_intergenic_TPM/ in at least 2 out of the three replicate libraries
- Not see any such expression in somatic libraries
- Not have a highly similar core-chromosome paralogue
 
Hence these scripts mainly sort the data to identify GRC-linked genes in this manner

First we run the script 01_get_expressed_genes.py. This script 
1) Generates a table of expressed genes only where expressed genes = genes with a TPM > intergenic_trheshold present in 2 or more samples (see outputs/GRC_gene_expression.tsv) 
2) Parses gene IDs for expressed GRC genes
3) Using GRC gene ID's parses out the gene sequences from B_cop GRC genome
4) Outputs three FASTA files for GRC genes expressed in soma, germ, and both (see /outputs/
5) Parses out all core genes from annotated B_cop genome
6) Produces combined fasta file contining all core chromosome genes

It requires the reference genome (-g), annotation (-a), the directory where files generated by /01_RNAseq_Mapping/ reside and the expression theshold set in /02_intergenic_TPM/. 

For this study it was run like so 
```
python 01_get_expressed_genes_for_BLAST.py -g ../Annotations/idBraCopr2.1.primary.masked -a ../Annotations/bcop_core_GRC.gff3 -o ../01_STAR_TPM/ -t 0.22 
```
We then BLAST the expressed GRC genes generated first against the annotated core genes, and then the full core genome. This is done to identify GRC-gene similarity to core-chromosome paralogues.

We BLAST against the core genes in 02_BLAST_GRCgenes.sh like so 
```
### STEP 1) BLAST GRC genes against core genes ###

# Make a blast data base of core-chromosome genes 
echo "making BLASTDB from core chromosome genes"
makeblastdb \
-in core_genes.fasta \
-dbtype nucl \
-parse_seqids \
-out core_genes_DB

# Blast GRC genes against core genes 
echo "BLASTn for GRC genes"
blastn \
-query GRC_genes_to_BLAST.fasta \
-db core_genes_DB \
-out GRC_gene_BLAST_output.tsv \
-outfmt '6 std qlen slen qseq sseq'
```
Then we BLAST the GRC-linked gene against the core genome (in case GRC-gene paralogues are not annotated correctly). We utilise a python script to mask out the GRC genome (located in /outputs/) 
```
### STEP 2) BLAST GRC genes against masked core genome ###

# Sync files in
# GRC genome
rsync -av /mnt/loki/ross/assemblies/flies/sciaridae/Bradysia_coprophila/idBraCopr2.1.primary.masked.fa $SCRATCH
# Masking script
rsync -av /mnt/loki/ross/flies/sciaridae/GRCs/GRC_expression/Bradysia_coprophila/03_BLAST/scripts/mask_genome.py $SCRATCH

# Make masked GRC genome
python3 mask_genome.py -g idBraCopr2.1.primary.masked.fa -o masked_GRC_genome.fasta -m GRC

# Make a blast data base of core-chromosome genes 
echo "making BLASTDB from maksed genome"
makeblastdb \
-in masked_GRC_genome.fasta \
-dbtype nucl \
-parse_seqids \
-out masked_GRC_genome_DB

# Blast GRC genes against whole genome 
echo "BLASTn for GRC genes"
blastn \
-query GRC_genes_to_BLAST.fasta \
-db masked_GRC_genome_DB \
-out GRC_genome_BLAST_output.tsv \
-outfmt '6 std qlen slen qseq sseq'
```
These BLAST results are then passed through another custom python script, 03_get_BLAST_table.py, in order to generate a tbale of expressed GRC genes and their respective best-hit core paralogue. 
```
python 03_get_BLAST_table.py -t 
```

