# 3. Generating expressed GRC data 

For a GRC-linked gene to be considered expressed in this study, it must satisfy a few conditions
- Have an TPM > cutoff generated in /02_intergenic_TPM/ in at least 2 out of the three replicate libraries
- Not see any such expression in somatic libraries
- Not have a highly similar core-chromosome paralogue
 
Hence these scripts mainly sort the data to identify GRC-linked genes in this manner

First we run the script 01_get_expressed_genes.py. This script 
1) Generates a table of expressed genes only where expressed genes = genes with a TPM > intergenic_trheshold present in 2 or more samples (see outputs/GRC_gene_expression.tsv) 
2) Generates nucleotide sequences for these genes in fasta format (see /outputs/fasta_files/GRC_genes.nucl.fasta)
3) Produces combined fasta file contining all core chromosome genes (not present; too large to fit on GitHub)

It requires the reference genome (-g), annotation (-a), the directory where files generated by /01_RNAseq_Mapping/ reside and the expression theshold set in /02_intergenic_TPM/. 

For this study it was run like so 
```
python 01_get_expressed_genes_for_BLAST.py -g ../Annotations/idBraCopr2.1.primary.masked -a ../Annotations/bcop_core_GRC.gff3 -o ../01_STAR_TPM/ -t 0.22 
```
We then run 02_BLAST_GRCgenes.sh to BLAST the expressed GRC-linked gene sequences we just generated first against the annotated core genes (generated in the previous step), and then the full core reference genome. This is done to identify GRC-gene similarity to core-chromosome paralogues.
```
### STEP 1) BLAST GRC genes against core genes ###

# Make a blast data base of core-chromosome genes 
echo "making BLASTDB from core chromosome genes"
makeblastdb \
-in core_genes.fasta \
-dbtype nucl \
-parse_seqids \
-out core_genes_DB

# Blast GRC genes against core genes 
echo "BLASTn for GRC genes"
blastn \
-query GRC_genes_to_BLAST.fasta \
-db core_genes_DB \
-out GRC_gene_BLAST_output.tsv \
-outfmt '6 std qlen slen qseq sseq'
```
When we BLAST the GRC-linked gene against the core genome (in case GRC-gene paralogues are not annotated correctly), we utilise a python script to mask out the GRC genome (located in /outputs/) 
```
### STEP 2) BLAST GRC genes against masked core genome ###

# Sync files in
# GRC genome
rsync -av /mnt/loki/ross/assemblies/flies/sciaridae/Bradysia_coprophila/idBraCopr2.1.primary.masked.fa $SCRATCH
# Masking script
rsync -av /mnt/loki/ross/flies/sciaridae/GRCs/GRC_expression/Bradysia_coprophila/03_BLAST/scripts/mask_genome.py $SCRATCH

# Make masked GRC genome
python3 mask_genome.py -g idBraCopr2.1.primary.masked.fa -o masked_GRC_genome.fasta -m GRC

# Make a blast data base of core-chromosome genes 
echo "making BLASTDB from maksed genome"
makeblastdb \
-in masked_GRC_genome.fasta \
-dbtype nucl \
-parse_seqids \
-out masked_GRC_genome_DB

# Blast GRC genes against whole genome 
echo "BLASTn for GRC genes"
blastn \
-query GRC_genes_to_BLAST.fasta \
-db masked_GRC_genome_DB \
-out GRC_genome_BLAST_output.tsv \
-outfmt '6 std qlen slen qseq sseq'
```
These BLAST results (located in outputs/GRC_v_Core_BLAST/) are then passed through another custom python script, 03_get_BLAST_table.py, in order to generate a table of expressed GRC genes and their respective best-hit core paralogue. This allows us to identify any GRC-linked genes with highly similar core paralogues which decreasue our condifence in their expression.

We pass the GRC_gene_expression.tsv generated by 01_get_expressed_genes_for_BLAST.py (-t), the GRC vs core-gene (-c) and the GRC vs core-genome outputs (-g) like so 
```
python 03_get_BLAST_table.py -t GRC_gene_expression.tsv -c GRC_v_Core_gene_BLAST_output - g GRC_v_Core_genome_BLAST_output.tsv
```
Our analysis now moves onto focusing on the confidently expressed GRC genes. 

First, we use a python script to generate translated amino-acid sequences for downsteam homology analysis. It requires the path to the output of the previous step (outputs/GRC_BLAST_table.tsv) (-t), along with the reference genome (-g) and annoation (-a).
```
python 04_get_GRC_proteins.py -t GRC_BLAST_table.tsv -g ../Annotations/idBraCopr2.1.primary.masked.fa -a ../Annotations/bcop_core_GRC.gff3
```
This script outputs a list of translated amino-acid sequences for all entries in GRC_BLAST_table.tsv (see /outputs/fasta_files/GRC_transcripts.fasta)

Note, some of these genes are not confidently GRC expressed due to mismapping/having a similar core paralogue. However, we still BLAST all these genes in the next steps, and manually curate our final list afterwards.

Using these amino-acid sequences, we use 05_BLAST_GRC_proteins.sh to BLASTp against the NCBI nr protein database
```
blastp \
-query /mnt/loki/ross/flies/sciaridae/GRCs/GRC_expression/Bradysia_coprophila/03_BLAST/outputs/GRC_transcripts.fasta \
-db /mnt/loki/db/ncbi_nr/nr \
-outfmt "6 std sscinames staxids stitle" \
-evalue 0.05 \
-num_threads 16 \
-out /mnt/loki/ross/flies/sciaridae/GRCs/GRC_expression/Bradysia_coprophila/03_BLAST/outputs/GRC_transcripts_BLAST_output.tsv
```
We then run InterPro Scan (online: https://www.ebi.ac.uk/interpro/search/sequence/) and parse the output using 06_get_interpro_summary.py to generate a more readable file (/outputs/GRC_gene_homology_BLAST/iprscan5_GRC_genes.tsv) 
```
python3 06_get_interpro_summary.py
```
We also tBLASTn GRC-gene sequences against repbase, to better idenitfy any of these genes are transposable/repetative elements 
```
tblastn \
-query /mnt/loki/ross/flies/sciaridae/GRCs/GRC_expression/Bradysia_coprophila/03_BLAST/outputs/GRC_transcripts.fasta \
-db /mnt/loki/db/Repbase25.09/Repbase25.09.fasta \
-outfmt "6 std" \
-evalue 1e-5 \
-num_threads 16 \
-out /mnt/loki/ross/flies/sciaridae/GRCs/GRC_expression/Bradysia_coprophila/04_all_GRCgene_BLAST/outputs/GRCtranscript_repbase_output.tsv
```
